crates/dbs-miniball/src/vm-vcpu/src/vcpu/mod.rs

```diff
@@ -0,0 +1,735 @@
+    VcpuSetVcpuEvents(kvm_ioctls::Error),
+
+    /// Failed to set KVM vcpu xcrs.
+    #[error("Failed to set KVM vcpu xcrs: {0}")]
+    VcpuSetXcrs(kvm_ioctls::Error),
+
+    /// Failed to set KVM vcpu xsave.
+    #[error("Failed to set KVM vcpu xsave: {0}")]
+    VcpuSetXsave(kvm_ioctls::Error),
+
+    /// Failed to set KVM vcpu reg.
+    #[error("Failed to set KVM vcpu reg: {0}")]
+    VcpuSetReg(kvm_ioctls::Error),
+
+    /// A call to cpuid instruction failed on x86_64.
+    #[error("failure while configuring CPUID for virtual CPU on x86_64")]
+    CpuId(dbs_arch::cpuid::Error),
+
+    /// The call to KVM_SET_CPUID2 failed on x86_64.
+    #[error("failure while calling KVM_SET_CPUID2 on x86_64")]
+    SetSupportedCpusFailed(#[source] kvm_ioctls::Error),
+
+    /// Error configuring the MSR registers on x86_64.
+    #[error("failure while configuring the MSR registers on x86_64")]
+    MSRSConfiguration(dbs_arch::regs::Error),
+
+    /// Error configuring the general purpose registers on x86_64.
+    #[error("failure while configuring the general purpose registers on x86_64")]
+    REGSConfiguration(dbs_arch::regs::Error),
+
+    /// Error configuring the special registers on x86_64.
+    #[error("failure while configuring the special registers on x86_64")]
+    SREGSConfiguration(dbs_arch::regs::Error),
+
+    /// Error configuring the page table on x86_64.
+    #[error("failure while configuring the page table on x86_64")]
+    PageTable(dbs_boot::Error),
+
+    /// Error configuring the floating point related registers on x86_64.
+    #[error("failure while configuring the floating point related registers on x86_64")]
+    FPUConfiguration(dbs_arch::regs::Error),
+
+    /// Cannot set the local interruption due to bad configuration on x86_64.
+    #[error("cannot set the local interruption due to bad configuration on x86_64")]
+    LocalIntConfiguration(dbs_arch::interrupts::Error),
+
+    /// Cannot create msr.
+    #[error("cannot create msr")]
+    CreateMsrs,
+}
+
+/// Dedicated Result type.
+pub type Result<T> = result::Result<T, Error>;
+
+#[derive(Clone)]
+pub struct VcpuConfig {
+    pub id: u8,
+    #[cfg(target_arch = "x86_64")]
+    pub cpuid: CpuId,
+    #[cfg(target_arch = "x86_64")]
+    // This is just a workaround so that we can get a list of MSRS.
+    // Just getting all the MSRS on a vcpu is not possible with KVM.
+    pub msrs: Msrs,
+}
+
+#[derive(Clone)]
+pub struct VcpuConfigList {
+    pub configs: Vec<VcpuConfig>,
+}
+
+impl VcpuConfigList {
+    /// Creates a default configuration list for vCPUs.
+    pub fn new(_kvm: &Kvm, num_vcpus: u8) -> Result<Self> {
+        if num_vcpus == 0 {
+            return Err(Error::VcpuNumber(num_vcpus));
+        }
+
+        #[cfg(target_arch = "x86_64")]
+        let base_cpuid = _kvm
+            .get_supported_cpuid(kvm_bindings::KVM_MAX_CPUID_ENTRIES)
+            .map_err(Error::KvmIoctl)?;
+
+        #[cfg(target_arch = "x86_64")]
+        let supported_msr_list =
```

Comment:
Because the `supported_guest_msrs()` only return `MsrList`. I can only implement the function to get `Msrs` here.

`supported_guest_msrs()` in dbs-arch:
```
pub fn supported_guest_msrs(kvm_fd: &Kvm) -> Result<MsrList> {
    let mut msr_list = kvm_fd
        .get_msr_index_list()
        .map_err(Error::GetSupportedModelSpecificRegisters)?;

    msr_list.retain(|msr_index| msr_should_serialize(*msr_index));

    Ok(msr_list)
}
```

`supported_guest_msrs()` in vmm-reference:
```
pub fn supported_guest_msrs(kvm_fd: &Kvm) -> Result<Msrs> {
    let mut msr_list = kvm_fd
        .get_msr_index_list()
        .map_err(Error::GetSupportedMSR)?;

    msr_list.retain(|msr_index| msr_should_serialize(*msr_index));

    let mut msrs =
        Msrs::new(msr_list.as_fam_struct_ref().nmsrs as usize).map_err(|_| Error::CreateMsrs)?;
    let indices = msr_list.as_slice();
    let msr_entries = msrs.as_mut_slice();
    // We created the msrs from the msr_list. If the size is not the same,
    // there is a fatal programming error.
    assert_eq!(indices.len(), msr_entries.len());
    for (pos, index) in indices.iter().enumerate() {
        msr_entries[pos].index = *index;
    }

    Ok(msrs)
}
```